### DISCUSSION TITLE
[Proposal] Native code/module access via WASI/VFS

### DISCUSSION THREAD

**Discussion by hugovincent** [veracruz-project/veracruz#237] [Ideas]
Status: work in progress.

Rationale:
----------

We want a way of exposing certain native functionality (such as side-channel safe
cryptography, or efficient NN inference) to Wasm programs on Veracruz.

The normal Wasm way of doing this is with imports, directly giving the Wasm
program functions it can call. Depending on the implementation of the Wasm runtime,
this may imply copying of data in and out of the Wasm linear memory. It also implies
giving the native functionality access to the Wasm Linear Memory (to read or write
any data not passed by value), which is a security risk for arbitrary native code.

While Wasm Interface Types [1] would be helpful here, for Veracruz we additionally
want to restrict (at least for now) the functionality that is made available with 
native modules, and for now don't want programs to be able to bring their own arbitrary
native modules. This is to limit the policy complexity, to maximise the benefit we
get from sandboxing in the first place, and to make our future lives easier as we
start looking at bounding program behaviour. 

Therefore, for now, we propose to add only necessary native functionality as trusted, 
Veracruz-provided modules, and to access them primarily using the WASI system interface
and VFS. The first such module will be for cryptography, where the functionality can
not adequately be implemented in Wasm due to the inability to write side-channel safe
code portably in Wasm (and also, where native instructions such as Armv8-CE or AES-NI
provide a very significant performance uplift). 

Patterns:
---------

1) Direct VFS access: native services are accessed using a "device" file on the VFS.
   This has the advantages of simplicity and familiarity, and disadvantages of
   requiring multiple copies of data in memory (e.g. to decrypt a file in the VFS, the
   program needs to read the ciphertext into linear memory, write it to the device file,
   and read back the results into linear memory). This pattern also avoids confused deputy
   threats (see below), since the native module can not directly accces the VFS. 

2) Indirect VFS access: native services are accessed using a device file as above, but
   these native services can also directly access the VFS. Instead of copying via the 
   linear memory, the program hands the native functionality a handle/capability to an
   object on the VFS (such as an encrypted blob) and another capability to
   a VFS object which will contain the result. We must use capabilities here to avoid
   confused deputy attacks, wherein a program can e.g. confuse the decryption module into
   giving it access to a VFS file it does not have policy-granted permission to access. 
   
As with POSIX and other programming of this sort, both approaches suffer from the challenge
of in-band vs out-of-band communications and signalling, e.g. for the decryption service
example, where the program supplies the decryption key. Traditionally this is solved
with `ioctl`s and `signal`s; we want to avoid that complexity if possible in our VFS,
however the alternatives (e.g. sockets) don't seem better. For the first prototype, we
propose to handle comms (e.g. for keying the decryptor module) in band using an RPC style
mechanism (e.g. commands and arguments serialized to protobuf and written to the device file)
and to ignore out of band signalling. 

Note that it's possible to combine these models, e.g. input indirectly, result directly, etc.

Example:
--------

Here is a very rough sketch of how a file decryption service might look (from the Wasm program's perspective):

1) Program opens `/crypto/file-decryption/aes` as `<cap-aes>`
2) Program creates a temporary file on the VFS for the decrypted plaintext,
   as `<cap-result>`. No one else will have a capability to access this. 
3) Program writes the following JSON-coded [2] string to `<cap-aes>`:
      ```JSON
      {
         "mode":   "cbc",
         "key":    "0x1234...",
         "source": <capability/FD of ciphertext file>,
         "dest":   <cap-result>,
      }
      ```
   and closes `<cap-aes>`. Conceptually, this triggers the operation itself, 
   perhaps blocking until it completes [3].
4) Program reads results from `<cap-result>` and goes on with it's day. 

Internally, the AES file decryption module is implemented as part of the Veracruz
binary, like the rest of the VFS – a dynamically extensible design (i.e. runtime loadable
native modules) is possible within this interface, but out of scope for this proposal.

(Note: some operations might benefit from in-place/destructive processing;
in such cases, the API could permit the same cap be passed as both source
and dest, although this is messy and complicates the capability system).

[1] https://github.com/WebAssembly/interface-types/blob/main/proposals/interface-types/Explainer.md

[2] JSON in strings avoids requiring the application to have a dependency on a library for
    object serialization (such as protobuf or Cap'n Proto) for the common case where the 
    application only has to write to (not read from) these synthetic files. Such strings can
    easily be constructed with standard-library functionality in most languages. If the
    application wants to, it can of course include a JSON library for constructing and parsing
    these strings.

[3] There are obvious problems and obvious solutions to this, we can deal with them later. A
    simple solution would be to have this operation return immediately, and have subsequent
    reads from `<cap-result>` block until the operation has produced sufficient result for the 
    read to return. If the application closes `<cap-source>` before reading from `<cap-result>`
    (and no other entities hold a capability to `<cap-source>`) then the system could
    transparently perform the operation destructively. This could be extended to chaining
    multiple native operations together (e.g. file decrypt and video decode) without having
    to roundtrip in and out of Wasm linear memory.

**Comment by ShaleXIONG** [comment]
Question: "Wasm execution blocks until the operation completes [3]." How does the wasm execution know when to block? In the sense that the wasm program might write the file, e.g. `/crypto/file-decryption/aes`, in several wasi_fd_write calls, (it really depends on the compiler) or never close the file `/crypto/file-decryption/aes`.

**Reply by ericvh** [reply]
write doesn't return until operation completes would be the simple form

**Reply by hugovincent** [reply]
The wasm program doesn't need to know... a few possible implementations:
* the VFS write call implementation passes the command string written to `<cap-aes>` before returning; the native module greedily parses the command and triggers the actual AES operation as soon as it sees a valid command. 
* the operation is triggered on the file close (this creates some challenges around error handling). 
* the subsequent read from `<cap-result>` blocks  until it has sufficient result to fill the  (here again, the VFS read implementation passes control to the AES module, and that doesn't return control to the VFS until ready).

**Reply by hugovincent** [reply]
(Also, I clarified the proposal text itself).

**Reply by ShaleXIONG** [reply]
@ericvh  If you are suggesting the Veracruz side automatically block, I do not think it will work very well e.g. `std::fs::write("foo.txt", some_content)` might be compiled, in my experience, to several wasi level fd_write calls in sequence, when do we know it is the last fd_write in its case.
If you are suggesting the wasm program explicit block and call into the AES module, what is the mechanism here?

**Reply by ShaleXIONG** [reply]
@hugovincent I think block on the subsequent read on `<cap-result>` is probably the best way. As we cannot guarantee `<cap-aes>` is closed from the wasm side.

**Reply by hugovincent** [reply]
The `fd_write` call should pass control to the native module before it returns. Since the command write could be fragmented into several `fd_write` calls, we can adopt one of the possibilities above, e.g. ~greedily parse, or wait til closure before parsing and performing the command,~ block the subsequent result read.

**Reply by ericvh** [reply]
> The wasm program doesn't need to know... a few possible implementations:
> 
> * the VFS write call implementation passes the command string written to `<cap-aes>` before returning; the native module greedily parses the command and triggers the actual AES operation as soon as it sees a valid command.
> * the operation is triggered on the file close (this creates some challenges around error handling).
> * the subsequent read from `<cap-result>` blocks  until it has sufficient result to fill the  (here again, the VFS read implementation passes control to the AES module, and that doesn't return control to the VFS until ready).

this is probably more for future work, but the only reason to not block return on write until operation completes is if we are doing multi-threading, in which case we may have to worry about multiple invocations.  The way we dealt with this in Plan 9 was using a construct called the clone file system which could be simplified in various ways depending on interaction model, but essentially the open would trigger creation of an instance represented by a subdirectory (or a sub-file), operations proceeded in that subdirectory and close deallocates it.  write/read blocking is used to handle completion events typically, although asynchronous operations can use an event file in the subdirectory to signal progress (but again, reads on those are basically blocking...so we'd need to understand our thread semantics)

**Reply by ericvh** [reply]
> @ericvh If you are suggesting the Veracruz side automatically block, I do not think it will work very well e.g. `std::fs::write("foo.txt", some_content)` might be compiled, in my experience, to several wasi level fd_write calls in sequence, when do we know it is the last fd_write in its case. If you are suggesting the wasm program explicit block and call into the AES module, what is the mechanism here?

ew...that's nasty.  What is the basis for it chopping up your write?

If that's the case its much more difficult and we'd need to buffer and then parse out the end of sequence (closing json bracket I suppose) at which point we'd block.  Doesn't change much, but would make code uglier.

**Reply by hugovincent** [reply]
I changed the proposal text accordingly.

**Reply by ericvh** [reply]
> The `fd_write` call should pass control to the native module before it returns. Since the command write could be fragmented into several `fd_write` calls, we can adopt one of the possibilities above, e.g. ~greedily parse, or wait til closure before parsing and performing the command,~ block the subsequent result read.

this could (and maybe should) be managed in the wasi filesystem interface layer -- essentially buffering the write until we detect the read at which point we pass the buffer to the aes function?

**Reply by hugovincent** [reply]
Good point about multithreading – but that doesn't exist in Wasm at this time, so I ignored it :-) 
Triggering the operation and blocking on whatever write call closes a complete command (i.e. closing `}` here) is probably the cleanest way (though still a bit nasty) of avoiding the hazard.

> ew...that's nasty. What is the basis for it chopping up your write?

The standard library (Rust `std`, `libc` etc) is free to chop up `fwrite()`/etc for whatever reason, e.g. for buffer management. Lots of systems (used to?) buffer small writes to optimize the common case of `printf()` by reducing the number of syscalls. AFAIK it shouldn't chop up a raw `write()` (here, `fd_write()`) call.

**Reply by ShaleXIONG** [reply]
@ericvh  I do not know the full story of why the rust-to-wasm compile decides to chop write. My guess here is:
- Rust-to-wasm compiler assumes the file system might run out of space, so they decide to write as many blocks as possible;
- The content might be located in non-continuous blocks, depending on how rust decide to implement basic types such as Vec and String.

To conclude I think we can block on the read of `<cap-result>`. It might not be applied to other modules, for example, ML module. However it is a starting point.

**Reply by hugovincent** [reply]
@ShaleXIONG are you seeing chopping on standard library `fwrite` (or equivalent) or on something akin to POSIX `write` itself? See https://stackoverflow.com/a/11414265

**Reply by ShaleXIONG** [reply]
@hugovincent  Some `fwrite` equivalent calls from the rust standard lib has been chopped, when compile to wasm.

**Comment by egrimley-arm** [comment]
"and receives back a capability to another object containing the result": Is that bit wrong/misleading? Because the JSON example suggests that the capability for the result is created by the caller and passed into the service.

**Reply by hugovincent** [reply]
Yeah, it's a simplification. I'll change the text to be more clear.

**Comment by ericvh** [comment]
I'd simplify to include source data in the request, this removes any requirement for the component program to have to access the wasi visible VFS.  The response includes the result.  Any sort of streaming (seek style operations) gets handled by the WASM accessing the interface through multiple invocations.  (which is pattern 1).  Pattern 2 will be a useful optimization, but maybe get the MVP first.

Pattern 1 makes the interface on the native module essentially equivalent to what one might expect from a web application, all data it needs to process the request is in the request body, and it sends the result.

**Reply by hugovincent** [reply]
Yeah, we considered that previously, and that can be done with the direct pattern I describe. It does imply that the cipher text has to be loaded into Wasm linear memory, only to be immediately written out again and never otherwise used; pattern 2 was an attempt to improve efficiency, but we could of course just start with the simpler direct model.

**Reply by ericvh** [reply]
I get that, and I think its useful, but what I think it ends up begging for is a deeper libos stack where WASI and the native module can both indirect through the lower filesystem implementation code --- OR --- we have some concept of shared memory blocks but that sounds like its probably more complicated due to the linear memory WASM problem.

**Reply by ericvh** [reply]
I think it makes sense to get an the simple version working first and then iterate given that this is gonna be a bottleneck for the IOTEX demo and at the end of the day while the extra copies will slow things down, the wasm ML is still gonna dominate the perf problem until we engineer that bit out.

**Reply by ericvh** [reply]
Thinking about it, in some ways, your passing around the capabilities is essentially passing around shared memory handles -- just in this case the shared memory is the buffer of the file pointed to by the capability which right now is an element pointed to by an array of capabilities in the ramfs.  So essentially you are giving handles to memory in the linear memory which may have its own issues given that the native programs might easily escape that range and cause havoc.  So -- if we go that route, what we are really talking about is keeping a capability referenced object store where capabilities reference objects and we store that outside of wasm linear memory so that all can access and share?

**Reply by hugovincent** [reply]
Yep I agree, simple is good, and we can extend later.

**Reply by hugovincent** [reply]
@ericvh I don't think it's shared memory handles... certainly not handles in the Wasm linear memory. The whole point of indirecting via the VFS is avoiding that (at the cost of at least some copies)... so not sure I follow you.

**Reply by ericvh** [reply]
The capabilities are pointing to buffers in the wasm linear memory, are they not? @ShaleXIONG ?  This is why passing a capability is essentially passing a reference to that IMHO.  Or is it just in WASI memory and therefore better?  I suppose given AES will just be a library call, the fs handling bits are going to be in WASI anyways so they can just pass the pointer to the ramfs buffer given the capability, but that gets back to my original concern about that opening the cookie jar to the library call (ie. we are all in the same process space so it could f'up whatever it wanted to as there is no page table protection)

**Reply by hugovincent** [reply]
Capabilities are file handles, nothing to do with the Wasm linear memory. WASI currently follows the Capsicum capability model, via CloudABI. These are not capabilities to memory (a la CHERI).

**Reply by ShaleXIONG** [reply]
@ericvh the capabilities are bound to file descriptors. The capabilities do not do anything to the wasm linear memory. To be more precise, The capabilities in an FD constraint what WASI-API the wasm program can calls, hence manipulating the underneath VFS. Also to clarify, WASI does not specify any memory space, I believe when you say wasi memory, you mean the memory space where VFS occupies.

**Reply by ericvh** [reply]
Yes, I meant the memory space where the VFS occupies -- my assumption is that this memory is allocated within the runtime, and the wasm linear address space is allocated in the same runtime.  I suppose that doesn't mean that it is allocated in the linear address space, but there's also probably no OS enforced separation between the two -- just that the wasm code can't access that buffer directly.   But the AES library code could, and by extension probably access the rest of the memory space (including the linear address space map).

I guess I need a better description of how the runtime is organizing and protecting this memory.  A block diagram might be useful.

**Reply by hugovincent** [reply]
@ericvh Rust, mostly.

**Reply by dominic-mulligan-arm** [reply]
Capabilities are referencing objects in the file system data structure in the isolate's RAM, not within the linear memory of the Wasm program (though as @ericvh surmised above, the two are basically both data structures in the isolate's RAM, but morally kept separate).

**Reply by ShaleXIONG** [reply]
@ericvh the entire Veracruz runtime (memory space) is split into the following few components (boxes represent memory space):
```
   +++++++                                 +++++++++++++++++++++++++++++++++
|    VFS   |      <----WASI---->      | WASM program + WASM execution engine |
   +++++++                  ^             ++++++++++++++++++++++++++++++++
                            |
public Client API for clients to read and write files in VFS
                            |
                            |
                     Clients
```
- WASM execution engine occupies some memory inside Veracruz runtime and uses it as the linear memory space for WASM program. Note that different WASM execution engines have their own way to actually tightly control and manage the memory space and they often just provide simple read memory and write memory functions. We do not think we are in a position to hack any execution engine so we can do some fancy memory tricks.
- VFS lives in a separate memory space. There is no restriction it must access via WASI, but we currently assume this is the case.
- WASI effect is a bunch of imported functions for the WASM program, but standardized (in a sense that the compile know how to call these functions). The WASM execution engine just passes parameters of the form of primitive types, e.g. 32-bit numbers, and pointers, which points to some memory space in linear memory. In this proposal the idea is to let AES module directly access VFS memory space, e,g. `<cap-source>`, but the WASM program need to pass some crucial information to AES module, e.g. `<cap-aes>`. However, the challenge here is how to ensure the AES module access the VFS in a secured way and also how the WASM program passes information with the help of VFS and standardized WASI, for example, through writing to a predefined special file, `<cap-aes>` here.
- "public API for clients to read and write files in VFS". They are basically read, write and append a file. When clients send some requests in the protocol buffer form, it eventually lands here. These client API will be converted to WASI function calls, for unifying the VFS access pattern and most important, access control. I think AES can (1) use these APIs to access the VFS, (2) they write their own layer to link to WASI API , OR (3) they directly access VFS bypassing WASI calling convention?

**Comment by hugovincent** [comment]
Open questions:
* How do we serialize (into strings) capabilities/file handles?
* How do we create temporary files for results? Perhaps our WASI implementation could support something akin to `mktemp` (see: https://github.com/WebAssembly/wasi-libc/issues/229).

**Reply by ericvh** [reply]
the temporary files could actually be a characteristic of the underlying synthetic versus something we have to explicitly create.

**Reply by hugovincent** [reply]
Yeah, that's an option – in the example, that would mean you always read the result from a well known location (e.g. `/crypto/file-decryption/aes/result`. It does create various hazards, and reduce usage possibilities though, and is kind of hacky IMO.

**Reply by ericvh** [reply]
see my comments in the other thread on multi-thread variants.  However, at the end of the day, capabilities->path so serialization is just a path maybe?  If you don't have derrived capability writes, the open will fail -- but I guess that begs the question of what capabilities does the library have (as specified differently from the application), which means we'll have to specify that in the policy.  Actually have a look at Inferno's SSL device reference in another comment -- it is essentially the same example (w/different encryption algos)

**Reply by hugovincent** [reply]
I disagree that a path is the same as a capability. A path is forgeable, and it's not a token of authority (rather the authorization check is done when the path is opened). 

> but I guess that begs the question of what capabilities does the library have (as specified differently from the application), which means we'll have to specify that in the policy

The native module, in the general case, should have no capabilities; any capabilities to access to the VFS should be passed to it; this is a fundamental part of the proposal. There may be specific cases where we want the native module to have ambient authority (i.e. have always-usable capabilities), or perhaps ambient authority to a private directory for that module, but can't think of a need in the general case.

The simple/obvious way of serializing capabilities is just to serialize the (integer) FD/handle; this requires the runtime to reverse-map back to the underlying VFS object in runtime memory. 

A nice property of something like `mktemp` (and sorry, I really meant more like [`tmpfile`](https://www.man7.org/linux/man-pages/man3/tmpfile.3.html)) in the capability context is it doesn't imply that a path ever even exists, it just creates a file and gives you a handle to it (and it's automatically destroyed on close).

**Reply by egrimley-arm** [reply]
`tmpfile` is exactly the sort of thing I was looking for and failing to find in WASI, though CloudABI seems to have something similar. (I couldn't even see how to implement `tmpfile` seeing as we don't have the POSIX-specified `/tmp` directory.) But if we're allowed to create new "syscalls", we can add something like `tmpfile`.

**Reply by hugovincent** [reply]
@egrimley-arm yep we don't have it since WASI doesn't have it. Since at the point we add it, we strictly speaking are diverging from WASI, we don't need to be WASI (or POSIX) compliant, so the lack of a `/tmp` is a non-issue. I would recommend we just implement it in a way that doesn't actually create a named file at all.

**Reply by ericvh** [reply]
The fundamental assumption here is that all native module file system calls have to have a path to the VFS that is identical to the wasi path.  So we'll need that library interface specified for whatever bindings we use for the native modules (I guess based on the verbal discussion we are only worried about Rust which probably simplifies things a great deal)

**Reply by dominic-mulligan-arm** [reply]
Hmm, it's hard to keep track of the current state of consensus (or lack of it) in this discussion, so apologies if I'm retreading old ground, here...  But, let's get much more concrete:

> The native module, in the general case, should have no capabilities; any capabilities to access to the VFS should be passed to it; this is a fundamental part of the proposal. There may be specific cases where we want the native module to have ambient authority (i.e. have always-usable capabilities), or perhaps ambient authority to a private directory for that module, but can't think of a need in the general case.

To my mind, as a first attempt, the native module should know nothing about the VFS at all (though I suppose this depends on where you draw the boundary around the "native module"), and should therefore never have a need to receive a capability.  We should try to hammer it into some restricted interface, e.g.:

```rust
trait Service {
  type Configuration;
  type Error;
  fn new() -> Self;
  fn configure(&mut self, config: Self::Configuration) -> Result<(), Self::Error>;
  fn serve(&self, in: &mut [u8]) -> Result<(), Self::Error>;
  fn about(&self) -> String;
}
```

or something equally straight-jacketed.  The runtime state then keeps track of the native services, and in particular we then have one of these registered with the runtime state that calls out to e.g. some `aes.c` code, or similar, when its methods are invoked.

On top of that, in the runtime, we build a layer that interposes between these services and the FS, reads appropriate configuration files and configures them, before calling out to them and writing results back.  This is triggered by the Wasm program reading from the results file (either a pre-existing file in e.g. `/native_srv/aes/`) or e.g. a temporary file created by a new system call analogue of `tempfile` (note: I don't think there's a strong argument that we should stay within the confines of `wasm32-wasi` if that's becoming a problem for us).

This has the advantage of:
1. Decoupling actually interfacing with services, and linking them into the runtime, from concerns about who reads what, and so on, from the file system.
2. Seems to be general enough to cover a large number of e.g. crypto operations, including encryption, signing, key generation, and similar, which are of interest in the first instance.

**Comment by ericvh** [comment]
Just to expand on my mount table derivation -- right now the wasi fs has a couple of structs managing the filesystem.  
At the top level you have:
* fd_table of pointers to a particular file table entry (are these capabilities?)
* inode_table of paths and their associated inode
* inode table - the actually buffers backing each file
* rights_table - permissions? (based on policy?)
* prestat_table - pre-open FD table (based on policy?)

So really at the file level, we are talking about inodes, my suggestion is we extend inode to have file system operation function table per inode.  For normal files (the ramfs) it just points to the current functions.  For synthetic files (such as aes) it points to the library implementations.  There can be a mount function which registers a library's synthetic path in the path_table and creates the inode with the right function hooks.

How to reference the library in the mount could be done one of two ways:
* predefined by policy (in which case it shows up as a preopen fd?) - we used namespace files for this which basically specified root filesystem of your process (which would include both the "root" and the synthetics).  It basically looks like a per-process fstab.
* mount system call exposed to the wasm program, since we will have a well-defined set of built ins you can use a canonical name for each.  In Inferno and Plan 9 we used a single character prefixed with a # -- for example, the SSL "device" in Inferno was #D (http://www.vitanuova.com/inferno/man/3/ssl.html) -- its also a good example of the clone file system from earlier.

**Reply by ShaleXIONG** [reply]
fd_table of pointers to a particular file table entry (are these capabilities?) <- YES, the capabilities are bound to FDs. fd_entry also includes other necessary information, e.g. where is the cursor etc.
rights_table - permissions? (based on policy?) <- YES, based on policy.
prestat_table - pre-open FD table (based on policy?) <- YES, based on policy.

There is a pending merge #193 , I already extend inode with file system operation, e.g. read and write files. We can further extend it to include synthetic files.

"How to reference the library in the mount" -> I think we predefine then by policy, similarly to "stdin" "stdout" and "stderr"?

**Reply by ericvh** [reply]
predefine works, but we'll still need a mapping method -- although I suppose if we have a table of native modules those could have a string to match against.

**Comment by egrimley-arm** [comment]
Some proposals:

1. There should be a single "device" file (`/dev/module`?) for accessing all native modules, rather than a different one for each module.

2. A native module must be invoked with a single call to `fd_write`. That way we don't need to worry about parsing the command to see if it's finished or commands from different libraries getting interleaved. We could provide library functions to facilitate making a single call to `fd_write` but it does not seem justified to complicate Veracruz code just so that applications can use standard library functions for writing to `/dev/module`.

3. The data written to `/dev/module` will consist of the target module's name, null-terminated, followed by the data for that module. If the name of the target module is not recognised then the call to `fd_write` will fail.

4. If the module name is recognised then the module is invoked and the call to `fd_write` will return with `Success` after the operation has completed. So the call to `fd_write` returns with `Success` even if the module encounters an error: that error would need to be communicated in some other module-dependent way; see below. (This does not preclude having a module with a command that sets off an asynchronous process that signals its completion by writing data to some other file, but in general there should be no need for the caller to do anything else, such close the file descriptor used for writing to `/dev/module` or read from some other file, to make the operation happen.)

5. The format of the data that follows the null-terminated module name is defined by the module. JSON would be a good default choice but it would be a bad idea to require JSON because some modules might require binary data, like a cryptographic key, and it would be daft to translate such things in and out of Base64 for no good reason.

6. Native modules can access the file system, but only through file descriptors provided by the caller. So a command to invoke a very simple AES decryption module might look something like: `"AESD\0{\"in\":\"123\",\"out\":\"456\",\"err\":\"789\",\"key\":...,}"`

7. Most modules will want to have a way of reporting an error, so one of the parameters should be a file descriptor of a file where an error message can be written. That means there is no way for the module to complain about the JSON being ill-formed, but a module that expects JSON could always provide a friendly response to `"MODULE\0{\"err\":\"123\"}"`. On the other hand, if the code is working and stable and the caller is not expecting there to be an error then there is no need for the caller to provide the `err` parameter.

**Reply by ericvh** [reply]
Spelling out in, out, and err here (#6) is probably less necessary -- in fact not just keeping it to strictly ordered operands is likely to create extra overhead unless we know we have lots of potentially optional arguments.  I'd also agree with your statement that json is probably overkill here since we are ultimately an API interface in the same sandbox on the same machine, formatting the input then parsing are just going to be extra overhead.

Having a single entry-point /dev/module has the negative side of not being able to spell out which native objects an application can access in the policy (well, at least the easiest version which is giving the app a capability handle to /dev/aes).  Ultimately having the string be parsed in the input stream versus in the file system is just changing where the filename parsing happens, but filesystem may make more sense because then things like discovery "just work" and we can leverage existing file system ENOTFOUND for modules not actually existing versus having to express those in the module error stream.

**Reply by egrimley-arm** [reply]
Yes, it's probably better to get the file system to parse the module name and map it to a function entry point and then also benefit from filesystem permissions. (That then makes it possible for the module to return a failure condition via the call to `fd_write` but I'm not sure we'd want to make use of that: it will probably be less confusing to have only the usual filesystem errors as return values from `fd_write`.)

**Comment by dreemkiller** [comment]
This looks well thought-out, butthere is a lot of string parsing in this proposal: The module name, JSON, etc. String parsing is extremely error-prone, and has a history of producing severe security vulnerabilities. Could we consider other options?

**Comment by dreemkiller** [comment]
This looks well thought-out, butthere is a lot of string parsing in this proposal: The module name, JSON, etc. String parsing is extremely error-prone, and has a history of producing severe security vulnerabilities. Could we consider other options?

**Reply by ericvh** [reply]
not to be difficult, but even if we disambiguate module inputs using the file system versus a string -- the parsing is still happening somewhere, just somewhere different.  Using JSON (or any other encoding to express arguments) could be less important, but at the end of the day parsing a binary stream versus a string...seems kinda like the same thing.  So it would seem you are arguing against serialization in general which likely buggers using the file system for the operational interface.
